keps/sig-scheduling/4328-node-affinity-based-eviction/README.md

```diff
@@ -0,0 +1,426 @@
+- Initial metrics are added
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Additional e2e tests are completed(if needed)
+- Additional metrics are added(if needed) 
+
+#### GA
+
+- No negative feedback
+- No bug issues reported
+
+### Upgrade / Downgrade Strategy
+
+#### Upgrade
+- Enable the feature gate in kube-apiserver, kube-controller-manager and kube-scheduler, and set `.spec.affinity.nodeAffinity.requiredDuringSchedulingRequiredDuringExecution` in pod.
+
+#### Downgrade
+
+- Disable the feature gate in kube-apiserver, kube-controller-manager and kube-scheduler, so that previously configured
+  `.spec.affinity.nodeAffinity.requiredDuringSchedulingRequiredDuringExecution` value will be ignored.
+- However, the `.spec.affinity.nodeAffinity.requiredDuringSchedulingRequiredDuringExecution` value of a Pod is preserved if it's previously configured; otherwise get silently dropped.
+
+### Version Skew Strategy
+
+kube-controller-manager, kube-scheduler and kube-apiserver will need to enable the feature gate for the full featureset
+to be present. 
+
+If kube-apiserver doesn't enable the feature gate, this new affinity will be silently dropped during pod creating, and will be silently dropped during pod updating if it's not been used.
+
+If kube-apiserver enable the feature gate, and:
+if only the kube-controller-manager enables the feature gate, the node affinity will be ignored during the scheduling phase.
+if only the kube-scheduler enables the feature gate, the node affinity will be ignored during the execution phase. 
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: NodeAffinityEviction
+  - Components depending on the feature gate: kube-controller-manager, kube-scheduler, kube-apiserver
+
+###### Does enabling the feature change any default behavior?
+
+No. It's a new API field, so no default behavior will be impacted.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+The feature can be disabled in Alpha and Beta versions by restarting kube-controller-manager, kube-scheduler and kube-apiserver with the feature-gate off. 
+After the feature-gate is disabled and kube-controller-manager is restarted, 
+`node-affinity-eviction` controller will also be disabled.
+In terms of Stable versions, users can choose to opt-out by not setting the corresponding field.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+The `RequiredDuringSchedulingRequiredDuringExecution` node affinity will take effect as expected.
+
+###### Are there any tests for feature enablement/disablement?
+
+Yes. Unit tests for this will be added to `pkg/registry/core/pod/strategy_test.go`.
```

Comment:
We also need unit/integration tests for the new controller, making sure that pods are not evicted when the feature is disabled.

Reply:
>  The e2e framework does not currently support enabling or disabling feature
gates. However, unit tests in each component dealing with managing data, created
with and without the feature, are necessary. At the very least, think about
conversion tests if API types are being modified.
> Additionally, for features that are introducing a new API field, unit tests that
are exercising the `switch` of feature gate itself (what happens if I disable a
feature gate after having objects written with the new field) are also critical.
You can take a look at one potential example of such test in:
https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282

Copied from template, I think the tests this question cares is the `switch` of the feature itself. 
The tests you said should belong to `Test Plan` part, and are added.

Reply:
Integration tests are not affected by the comment above. Only E2E tests.